### PR TITLE
Plugins fix

### PR DESCRIPTION
**Pull request by zrthxn**
### Summary of Changes
A list of all changes is given here. Screenshots are provided in comments to give temporal relation to commits.

Plugins List
- Group plugins by version, dont show a new entry for every version
- Create plugin view, remove plugin preview
- Implement pagination in plugin list

Dashboard
- Fix Table layout and reduce complexity 
- Improve UI of plugin list on dashboard

Create plugin form
- Improve UI of form
- Check if name is unique while typing
- Fix issues with file upload component

Plugin body
- Improve UI and layout
- Fix issues caused by removal of PF3

Screenshot (83)
Screenshot from 2021-06-05 15-32-58

### Problems Remaining
- API does not return plugin URL for PluginMetas, thus install button has no URL.
- Versions are not fetched in call for PluginMetas. This versions content is empty.

Closes: #203 
Closes: #101 
Closes: #197
Fixes: #198  
Closes: #200

### CODE REVIEW

**Comment by zrthxn** [comment]
Screenshot (86)

**Comment by zrthxn** [comment]
Screenshot from 2021-06-05 00-58-23
Screenshot from 2021-06-05 01-01-27

**Comment by zrthxn** [comment]
Screenshot from 2021-06-05 15-32-58

**Comment by zrthxn** [comment]
Rebased to include changes from [refactoring](https://github.com/zrthxn/ChRIS_store_ui/tree/refactoring) branch

**Comment by zrthxn** [comment]
Rebased to include changes from [refactoring](https://github.com/zrthxn/ChRIS_store_ui/tree/refactoring) branch

File: src/components/Plugins/components/PluginItem/PluginItem.jsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    return <StarIcon className={className} onClick={onStarClicked} />;
+  }
+
+  function onStarClicked() {
+    rest.onStarClicked();
+    if (isFavorite)
+      return
+
+    setStarActive(true);
```
Comment (jennydaman) [inline src/components/Plugins/components/PluginItem/PluginItem.jsx:33]: Mixed prop/state manipulation.

Whether the star should be grey or yellow should be determined by either prop alone, there should be no internal manipulation of state for something that is decided by a prop. If it is desirable to have these lines of code, they should be moved to `Plugins.jsx`

That is, the only purpose of `onStarClicked` should be to either:

- toggle `isFavorite` (where the design decision would be, stars are a toggle-type checkbox-like component
- call `rest.onStarClicked()` (where `Plugins.jsx` is responsible for providing the prop `onStarClicked={() => setIsFavorite(!isFavorite)}`

File: src/components/general/CopyURLButton.jsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import React, { useState } from "react";
+import { Button, Popover } from "@patternfly/react-core";
+import { DownloadIcon, CopyIcon } from "@patternfly/react-icons";
+import { CopyToClipboard } from "react-copy-to-clipboard";
```
Comment (jennydaman) [inline src/components/general/CopyURLButton.jsx:4]: Could you look into https://www.patternfly.org/v4/components/clipboard-copy/

**Review by jennydaman — CHANGES_REQUESTED** [review]

File: src/components/Dashboard/Dashboard.jsx [hunk 3]
```diff
@@ -30,28 +32,31 @@ class Dashboard extends Component {
       console.error(err);
     });
   }
-  showNotifications = (error) => {
+
+  initialize() {
```
[the remark this answers is not in the document]
Comment (zrthxn) [reply src/components/Dashboard/Dashboard.jsx:36]: No, not as such.

File: src/components/Plugin/Plugin.jsx [hunk 4]
```diff
@@ -274,6 +272,24 @@ export class Plugin extends Component {
         <div className={`plugin ${this.props.className}`}>
           {container}
         </div>
+
```
[the remark this answers is not in the document]
Comment (zrthxn) [reply src/components/Plugin/Plugin.jsx:275]: Do you mean the commented out bits?

File: src/components/Dashboard/components/DashPluginCardView/DashPluginCardView.jsx [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+      showEditConfirmation: true,
+      pluginToEdit: plugin
+    });
+  }
+  
+  handlePublicRepo(value) {
+    this.setState({ publicRepo: value });
+  }
+  
+  toggle = (value, id) => {
```
[the remark this answers is not in the document]
Comment (zrthxn) [reply src/components/Dashboard/components/DashPluginCardView/DashPluginCardView.jsx:123]: Alright will do.

File: src/components/general/CopyURLButton.jsx (hunk 2 above)
Comment (zrthxn) [reply src/components/general/CopyURLButton.jsx:4]: Deleting this single use component.

File: src/components/Developers/components/CopyToClipboard/CopyToClipboard.jsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import Clipboard from 'react-clipboard.js';
+import { Tooltip } from '@patternfly/react-core';
+import { CopyIcon } from '@patternfly/react-icons';
+import './CopyToClipboard.css';
+
+const CopyToClipboard = ({ clipboardText }) => (
```
[the remark this answers is not in the document]
Comment (zrthxn) [reply src/components/Developers/components/CopyToClipboard/CopyToClipboard.jsx:8]: Replacing.

File: src/components/Plugins/components/PluginItem/PluginItem.jsx (hunk 1 above)
Comment (zrthxn) [reply src/components/Plugins/components/PluginItem/PluginItem.jsx:33]: The purpose was to make the tooltip disappear after the star was clicked. But I think we can do without that.

File: src/components/Plugin/Plugin.jsx [hunk 7]
```diff
@@ -50,12 +50,16 @@ export class Plugin extends Component {
   async componentDidMount() {
     let pluginData;
 
+    console.log(this.state.pluginData)
```
Comment (jennydaman) [inline src/components/Plugin/Plugin.jsx:53]: Delete

**Review by jennydaman — CHANGES_REQUESTED** [review]

File: src/components/Plugin/components/PluginBody/PluginBody.jsx [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+import { sanitize } from 'dompurify';
+
+import './PluginBody.css';
+import CopyURLButton  from '../../../general/CopyURLButton';
+import Notification from '../../../Notification';
+import HttpApiCallError from '../../../../errors/HttpApiCallError';
+import { GithubAPIRepoError,GithubAPIProfileError, GithubAPIReadmeError } from '../../../../errors/GithubError';
+
+
+const removeEmail = (authors) => Array.isArray(authors) ?
```
[the remark this answers is not in the document]
Comment (zrthxn) [reply src/components/Plugin/components/PluginBody/PluginBody.jsx:25]: Yes iirc there is in some places, like mapping the string to an array to make the list of authors.
But this was also put in as a measure to accommodate arrays, when we eventually fix the authors list being a string of all authors to an array.

File: src/components/Plugin/Plugin.jsx (hunk 7 above)
Comment (zrthxn) [reply src/components/Plugin/Plugin.jsx:53]: Sorry, will do

**Pull request closed without merging**